server/src/main/java/org/elasticsearch/index/analysis/ElasticSynonymParser.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.analysis;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.synonym.SolrSynonymParser;
+import org.apache.lucene.util.CharsRef;
+import org.apache.lucene.util.CharsRefBuilder;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.io.IOException;
+
+public class ElasticSynonymParser extends SolrSynonymParser {
+
+    private boolean lenient;
+    private final Logger logger;
+
+    public ElasticSynonymParser(boolean dedup, boolean expand, boolean lenient, Analyzer analyzer) {
+        super(dedup, expand, analyzer);
+        this.lenient = lenient;
+        logger = Loggers.getLogger(getClass(), "ElasticSynonymParser");
+    }
+
+    @Override
+    public void add(CharsRef input, CharsRef output, boolean includeOrig) {
+        if (!lenient || (input.length > 0 && output.length > 0)) {
```

Comment:
`!lenient` we usually write as `lenient == false`

Reply:
@sohaibiftikhar  I am not sure I understand this condition `(!lenient || (input.length > 0 && output.length > 0)`,  can you please explain what you are trying to achieve here.

Reply:
So if `lenient` is not set we should proceed as before. For the case when lenient is set, if there are exceptions during `analyze` we return a zero-length `CharsRef`. When these synonym mappings are added using the `add` method we skip the ones that were left empty (`super.add` would throw exceptions otherwise). Does this make sense?
I believe if the condition was only `input.length > 0 && output.length>0` it would still work cause it never really reaches this part with zero length `input` or `output` with the non-lenient setting since it would throw an exception during analyze. But I just wanted to be sure that I only change the logic in case when `lenient` is set.

Reply:
@sohaibiftikhar Thanks for the explanation, it does makes sense and the logic is correct, but your condition by itself is not obvious.  I would either provide more extensive explanation before this condition, or I would instead override `addInternal` method of `SolrSynonymParser` parser where the logic can be obvious.

Reply:
I think overriding addInternal may not be the best idea because the method is private and it would mean repetition of the logic (it calls some other private methods that we'd need to copy as well). I will add the explanation above the condition as you suggested.